### PR TITLE
Add arrow-adbc packages

### PR DESCRIPTION
**Pull request by lidavidm**
Announced by the Apache Arrow project here: https://lists.apache.org/thread/lkplzc80gkghdjt30zrtp4spyrmbdpcz

ADBC is an API for database access for the Apache Arrow ecosystem. This PR adds the 0.1.0 release, including these packages:
- A driver manager for C/C++ and Python to dynamically load ADBC drivers (analogously to ODBC, JDBC),
- Drivers for PostgreSQL and SQLite.

Future releases will add more functionality. 

Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [x] If static libraries are linked in, the license of the static library is packaged.
- [x] Package does not ship static libraries. If static libraries are needed, [follow CFEP-18](https://github.com/conda-forge/cfep/blob/main/cfep-18.md).
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [x] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

### CODE REVIEW

**Comment by lidavidm** [comment]
CC @mariusvniekerk, how does this look? Probably I'll have to tweak it once it lands to get the arm64 builds working

**Comment by lidavidm** [comment]
@conda-forge/help-python-c, ready for review!

File: recipes/arrow-adbc/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+  - name: adbc-driver-manager
+    script: build-python.sh                                                      # [not win]
+    script: build-python.bat                                                     # [win]
+    requirements:
+      build:
+        - {{ compiler('cxx') }}
+        - cross-python_{{ target_platform }}                                     # [build_platform != target_platform]
+        - cython
+        - pip
+        - poetry
```
Comment (jorisvandenbossche) [inline recipes/arrow-adbc/meta.yaml:112]: poetry is not used anymore nowadays and can be removed?

Comment (lidavidm) [reply recipes/arrow-adbc/meta.yaml:112]: Thanks for catching that. Removed.

File: recipes/arrow-adbc/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+  sha256: a50f3dac985883042152938bdd69c843708fd50508e0f15ba3a179619c11f7b9
+  patches:
+    # Backport https://github.com/apache/arrow-adbc/issues/304, drop in 0.2.0
+    - patches/0001-fix-c-driver-postgresql-define-ntohll-etc-for-macOS.patch
+
+build:
+  number: 0
+
+outputs:
+  - name: adbc-driver-manager-cpp
```
Comment (carterbox) [inline recipes/arrow-adbc/meta.yaml:18]: Outputs with headers and shared libraries should have run_exports which pin an ABI compatible version of the output.

File: recipes/arrow-adbc/meta.yaml (hunk 1 above)
Comment (carterbox) [reply recipes/arrow-adbc/meta.yaml:112]: Poetry is still listed here...

File: recipes/arrow-adbc/build-python.sh [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    export ADBC_SQLITE_LIBRARY=$PREFIX/lib/libadbc_driver_sqlite.$LIB_SUFFIX
+else
+    echo "Unknown package ${PKG_NAME}"
+    exit 1
+fi
+
+export SETUPTOOLS_SCM_PRETEND_VERSION=$PKG_VERSION
+
+echo "==== INSTALL ${PKG_NAME}"
+$PYTHON -m pip install . -vvv --no-deps
```
Comment (carterbox) [inline recipes/arrow-adbc/build-python.sh:27]:
```suggestion
$PYTHON -m pip install . -vvv --no-deps --no-build-isolation
```

File: recipes/arrow-adbc/build-python.bat [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+)
+echo Unknown package %PKG_NAME%
+exit 1
+
+:BUILD
+
+set SETUPTOOLS_SCM_PRETEND_VERSION=%PKG_VERSION%
+
+echo "==== INSTALL %PKG_NAME%"
+%PYTHON% -m pip install . -vvv --no-deps || exit /B 1
```
Comment (carterbox) [inline recipes/arrow-adbc/build-python.bat:23]:
```suggestion
%PYTHON% -m pip install . -vvv --no-deps --no-build-isolation || exit /B 1
```

File: recipes/arrow-adbc/meta.yaml [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+      run:
+    test:
+      commands:
+        - test -f $PREFIX/include/adbc.h                                         # [unix]
+        - test -f $PREFIX/include/adbc_driver_manager.h                          # [unix]
+        - test -d $PREFIX/lib/cmake/AdbcDriverManager/                           # [unix]
+        - test -f $PREFIX/lib/pkgconfig/adbc-driver-manager.pc                   # [unix]
+        - test ! -f $PREFIX/lib/libadbc_driver_manager.a                         # [unix]
+        - test -f $PREFIX/lib/libadbc_driver_manager.so                          # [linux]
+        - test -f $PREFIX/lib/libadbc_driver_manager.dylib                       # [osx]
```
Comment (carterbox) [inline recipes/arrow-adbc/meta.yaml:36]: I suggest checking for a specific major SO name in order to guard against ABI breakages.

File: recipes/arrow-adbc/meta.yaml [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+        - if not exist %LIBRARY_LIB%\cmake\AdbcDriverPostgreSQL exit 1           # [win]
+        - if not exist %LIBRARY_LIB%\pkgconfig\adbc-driver-postgresql.pc exit 1  # [win]
+
+  - name: adbc-driver-sqlite-cpp
+    script: build-cpp.sh                                                         # [not win]
+    script: build-cpp.bat                                                        # [win]
+    requirements:
+      build:
+        - {{ compiler('c') }}
+        - {{ compiler('cxx') }}
```
Comment (carterbox) [inline recipes/arrow-adbc/meta.yaml:80]: There are linker warnings that `libstdcxx` is unused by the sqlite library. Either remove this compiler or add `ignore_run_exports_from: {{ compiler('cxx') }}`.

File: recipes/arrow-adbc/meta.yaml [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+        - adbc_driver_manager._lib
+
+  - name: adbc-driver-postgresql
+    script: build-python.sh                                                      # [not win]
+    script: build-python.bat                                                     # [win]
+    requirements:
+      build:
+        # We don't use these, but Conda needs them to know that this is a binary package.
+        - {{ compiler('c') }}
+        - {{ compiler('cxx') }}
```
[the remark this answers is not in the document]
Comment (carterbox) [reply recipes/arrow-adbc/meta.yaml:135]: This is OK as long as these shared objects are being built here with our build tools and not elsewhere.

File: recipes/arrow-adbc/meta.yaml [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+        - test -f $PREFIX/lib/libadbc_driver_manager.dylib                       # [osx]
+
+        - if not exist %LIBRARY_BIN%\adbc_driver_manager.dll exit 1              # [win]
+        - if not exist %LIBRARY_INC%\adbc.h exit 1                               # [win]
+        - if not exist %LIBRARY_INC%\adbc_driver_manager.h exit 1                # [win]
+        - if not exist %LIBRARY_LIB%\adbc_driver_manager.lib exit 1              # [win]
+        - if not exist %LIBRARY_LIB%\cmake\AdbcDriverManager exit 1              # [win]
+        - if not exist %LIBRARY_LIB%\pkgconfig\adbc-driver-manager.pc exit 1     # [win]
+
+  - name: adbc-driver-postgresql-cpp
```
Comment (carterbox) [inline recipes/arrow-adbc/meta.yaml:45]: The name of these cpp packages should be `libadbc-driver-postgresql`

**Review by carterbox — CHANGES_REQUESTED** [review]

**Comment by carterbox** [comment]
I still need to review the actual build artifacts, but here's some preliminary feedback.

File: recipes/arrow-adbc/meta.yaml [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+        - test -f $PREFIX/lib/libadbc_driver_sqlite.so                           # [linux]
+        - test -f $PREFIX/lib/libadbc_driver_sqlite.dylib                        # [osx]
+
+        - if exist %LIBRARY_INC%\adbc.h exit 1                                   # [win]
+        - if not exist %LIBRARY_BIN%\adbc_driver_sqlite.dll exit 1               # [win]
+        - if not exist %LIBRARY_LIB%\adbc_driver_sqlite.lib exit 1               # [win]
+        - if not exist %LIBRARY_LIB%\cmake\AdbcDriverSQLite exit 1               # [win]
+        - if not exist %LIBRARY_LIB%\pkgconfig\adbc-driver-sqlite.pc exit 1      # [win]
+
+  - name: adbc-driver-manager
```
Comment (carterbox) [inline recipes/arrow-adbc/meta.yaml:103]: adbc-driver-manager is shipping headers and c source code which are not needed

File: recipes/arrow-adbc/build-python.sh [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+    export ADBC_SQLITE_LIBRARY=$PREFIX/lib/libadbc_driver_sqlite.$LIB_SUFFIX
+else
+    echo "Unknown package ${PKG_NAME}"
+    exit 1
+fi
+
+export SETUPTOOLS_SCM_PRETEND_VERSION=$PKG_VERSION
+
+echo "==== INSTALL ${PKG_NAME}"
+$PYTHON -m pip install . -vvv --no-deps --no-build-isolation
```
Comment (carterbox) [inline recipes/arrow-adbc/build-python.sh:27]: The python packages are being built multiple times for python=3.11 only. Although the build matrix claims they are being built for various 3.8, 3.9, 3.10, 3.11. Not sure why that is.

**Review by carterbox — CHANGES_REQUESTED** [review]

**Comment by lidavidm** [comment]
Thanks for the review, I'll fix these when I get a chance.

Note that we're coming up on 0.2.0 which has a number of improvements which I'll sync back here (in particular, the Python packages can become noarch and the workaround isn't needed).

**Comment by lidavidm** [comment]
I've punted on macOS and Windows since a bunch of linking issues came up that I can't replicate on VMs/M1. So for now this package will have to be Linux only.

- Synced with upstream recipe (though now this recipe is ahead of upstream)
- Renamed packages
- Removed source/headers from Python adbc-driver-manager
- Added run-exports
- Updated checks to test for specific soname

I'm not sure what's going on with Python. One thing I was considering was one feedstock for packages with native libraries and one for the pure-Python bits that depend on those native libraries.

**Review by mariusvniekerk — APPROVED** [review]

**Comment by jorisvandenbossche** [comment]
I suppose it should be possible to avoid running the MacOS/Windows CI at all? (currently they take almost the same time as linux to set up some build envs, but still only ~10min, so not a big deal)

Otherwise looks good based on my limited conda-forge experience.

**Comment by lidavidm** [comment]
Once there's a feedstock, we can disable osx/windows in conda-forge.yml

**Pull request merged**